### DISCUSSION TITLE
How to setup connection to dev database?

### DISCUSSION THREAD

**Discussion by naXa777** [danielschuster-muc/potter-db#176] [Q&A] [answered]
## Problem

I'm following steps in the backend/README.md file and it seems that the essential part about connecting to database is missing.

when I execute one of the following commands:

```shell
rails db:create db:migrate
```

```shell
rails db:seed
```

```shell
rspec spec
```

I get error like
```
rails aborted!
ActiveRecord::ConnectionNotEstablished: connection to server at "localhost" (::1), port 5432 failed: Connection refused (0x0000274D/10061)
        Is the server running on that host and accepting TCP/IP connections?
```

## What have I tried to fix this error myself?

1. I've searched the source code for DB connection config.
    And I've discovered `database.yml`

2. Started PostgreSQL db in a Docker container. My Dockerfile looks like:
    ```
    FROM library/postgres
    ENV POSTGRES_USER potter_db_api
    ENV POSTGRES_PASSWORD any_password
    ENV POSTGRES_DB potter_db_api_development
    ```

3. Added `POTTER_DB_API_DATABASE_PASSWORD` env variable.

**Comment by danielschuster-muc** [comment]
To setup the dev database you have to install **PostgreSQL 14** on your local machine.

### Windows
Download Postgres from [here](https://www.postgresql.org/download/) and follow the setup wizard. 

### MacOS
```bash
brew install postgresql
brew services start postgresql
```

### Linux
```bash
sudo apt install -y postgresql postgresql-contrib libpq-dev build-essential
sudo -u postgres psql --command "CREATE ROLE \"`MY_USERNAME`\" LOGIN createdb superuser;"
```


Maybe it also works with docker, but I've never tried it.

**Comment by naXa777** [comment, marked as answer]
Oh man... it was my fault. I didn't check that my db is accepting connections. I needed to expose 5432 port correctly.

Also, I've added these missing properties to development and test environments in the database.yml file:

```yml
  username: potter_db_api
  password: <%= ENV["POTTER_DB_API_DATABASE_PASSWORD"] %>
```

without this change, when running `rails db:create db:migrate`, I've got this error:

```
connection to server at "localhost" (::1), port 5432 failed: fe_sendauth: no password supplied
Couldn't create 'potter_db_api_development' database. Please check your configuration.
rails aborted!
```

And with this change, the db migrations run successfully.
Thank you for your help and let's close the discussion.
